### PR TITLE
Closes #3310: browser-engine-gecko-nightly: Automatically recover and restore if content process is killed.

### LINKED ISSUE #3310: Recover when content process gets killed
Fenix issue: https://github.com/mozilla-mobile/fenix/issues/1577
GeckoView bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1557096

Currently we see `onCrash()` getting called for a `GeckoSession` - but we have no clue whether that is a crash or whether the content process just got killed by Android to reclaim some resources.

We can handle crashes (via `lib-crash`) and allow the user to restore manually - but we need a way to know whether a call to `onCrash()` was just for a process kill so that we can auto-restore.

### PR DESCRIPTION
**Pull request by pocmo**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

**Comment by pocmo** [comment]
Okay, then I'll make TC happy now! 🍸

**Comment by pocmo** [comment]
TC and I had a pedicure and massage at a local spa (💆‍♂ ). We then visited the zoo (🦁). And then we ended the day at a bar (🍻). Pretty great day. I think TC is happy now!

**Comment by Amejia481** [comment]
Nice!!!!
I need a day like that!

File: components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 1]
```diff
@@ -494,10 +494,33 @@ class GeckoEngineSession(
         override fun onCrash(session: GeckoSession) {
             stateBeforeCrash = lastSessionState
 
+            recoverGeckoSession()
+
+            notifyObservers { onCrash() }
+        }
+
+        override fun onKill(session: GeckoSession) {
+            // The content process of this session got killed (resources reclaimed by Androdi).
```
Comment (Amejia481) [inline components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: nit: Androdi -> Android

File: components/browser/engine-gecko-beta/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineViewTest.kt [hunk 2]
```diff
@@ -104,4 +105,24 @@ class GeckoEngineViewTest {
...
+        val engineSession = mock<GeckoEngineSession>()
+        val geckoSession = mock<GeckoSession>()
+        val geckoView = mock<NestedGeckoView>()
+
+        whenever(engineSession.geckoSession).thenReturn(geckoSession)
+        engineView.currentGeckoView = geckoView
+
+        engineView.render(engineSession)
+
+        Mockito.reset(geckoView)
```
Comment (Amejia481) [inline components/browser/engine-gecko-beta/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineViewTest.kt]: I'm not completely sure we want to reset the mock here?

Comment (pocmo) [reply components/browser/engine-gecko-beta/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineViewTest.kt]: The initial render call will already call `setSession()` and I want to just ignore that and start fresh to only verify what happens when calling `onProcessKilled()`.

**Comment by pocmo** [comment]
🚫 DO NOT LAND: The new GeckoView API was just backed out because of regressions :(
https://bugzilla.mozilla.org/show_bug.cgi?id=1557096

File: components/browser/engine-gecko-beta/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineViewTest.kt (hunk 2 above)
Comment (Amejia481) [reply components/browser/engine-gecko-beta/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineViewTest.kt]: 👍

**Comment by pocmo** [comment]
The patch landed in GV again. This should be unblocked after a rebase.

**Comment by pocmo** [comment]
Landed as cf82f8e2b7f542fa235000a80667eaf9203dfedf

**Pull request closed without merging**